src/main/java/appeng/parts/misc/ItemHandlerAdapter.java

```diff
@@ -0,0 +1,370 @@
+				// We have a chance to use this slot for injection
+				remaining = itemHandler.insertItem( i, remaining, true );
+				if( remaining == null )
+				{
+					break; // Awesome, full stack consumed
+				}
+			}
+		}
+		else
+		{
+
+			// For actually inserting the stack, we try to fill up existing stacks in the inventory, and then move onto free slots
+			// Think about this: In a storage drawer or barrel setup, we'd want to attempt to fill up containers that already have the
+			// item we're trying to insert first. The second phase should not cost a considerable amount of time, since it will just search
+			// for empty slots.
+			int slotCount = itemHandler.getSlots();
+
+			// This array is used to remember which slots were viable but skipped in the first phase
+			// This avoids calling getStackInSlot for each slot a second time. Hopefully the JVM will decide
+			// to allocate this array on the stack
+			boolean[] retry = new boolean[slotCount];
+
+			for( int i = 0; i < slotCount; i++ )
+			{
+				ItemStack stackInSlot = itemHandler.getStackInSlot( i );
+
+				if( stackInSlot == null )
+				{
+					retry[i] = true;
+					continue; // In the first phase, we try to top up existing item stacks
+				}
+
+				remaining = itemHandler.insertItem( i, remaining, false );
+				if( remaining == null )
+				{
+					break; // Awesome, full stack consumed
+				}
+			}
+
+			// If we reached this point, we still have items to insert, our first pass failed.
+			// Now we try to insert into empty slots
+			for( int i = 0; i < slotCount; i++ )
+			{
+				if( retry[i] )
+				{
+					remaining = itemHandler.insertItem( i, remaining, false );
+					if( remaining == null )
+					{
+						break; // Awesome, full stack consumed
+					}
+				}
+			}
+		}
+
+		// At this point, we still have some items left...
+		if ( remaining == orgInput ) {
+			// The stack remained unmodified, target inventory is full
+			return iox;
+		}
+
+		if( type == Actionable.MODULATE )
+		{
+			this.onTick();
+		}
+
+		return AEItemStack.create( remaining );
+
+	}
+
+	@Override
+	public IAEItemStack extractItems( IAEItemStack request, Actionable mode, BaseActionSource src )
+	{
+
+		ItemStack req = request.getItemStack();
+		int remainingSize = req.stackSize;
+
+		// Use this to gather the requested items
+		ItemStack gathered = null;
+
+		final boolean simulate = ( mode == Actionable.SIMULATE );
+
+		for( int i = 0; i < itemHandler.getSlots(); i++ )
+		{
+			ItemStack sub = itemHandler.getStackInSlot( i );
+
+			if( !Platform.isSameItem( sub, req ) )
+			{
+				continue;
+			}
+
+			ItemStack extracted;
+
+			// We have to loop here because according to the docs, the handler shouldn't return a stack with size > maxSize, even if we
+			// request more. So even if it returns a valid stack, it might have more stuff.
+			do
+			{
+				extracted = itemHandler.extractItem( i, remainingSize, simulate );
+				if( extracted != null )
+				{
+					if( extracted.stackSize > remainingSize )
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) Refactor this code to not nest more than 3 if/for/while/switch/try statements. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.orod.org/coding_rules#rule_key=squid%3AS134)